clippy_lints/src/unnecessary_indexing.rs

```diff
@@ -0,0 +1,159 @@
+    /// This code is unnecessarily complicated and can instead be simplified to the use of an
+    /// if..let expression which accessed the first element of the sequence.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let a: &[i32] = &[1];
+    /// if !a.is_empty() {
+    ///    let b = a[0];
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let a: &[i32] = &[1];
+    /// if let Some(b) = a.first() {
+    ///
+    /// }
+    /// ```
+    #[clippy::version = "1.78.0"]
+    pub UNNECESSARY_INDEXING,
+    complexity,
+    "unnecessary use of `seq.is_empty()` in a conditional when if..let is more appropriate"
+}
+
+declare_lint_pass!(UnnecessaryIndexing => [UNNECESSARY_INDEXING]);
+
+impl LateLintPass<'_> for UnnecessaryIndexing {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ rustc_hir::Expr<'_>) {
+        if let Some(if_expr) = clippy_utils::higher::If::hir(expr)
+            // check for negation
+            && let ExprKind::Unary(UnOp::Not, unary_inner) = if_expr.cond.kind
+            // check for call of is_empty
+            && let ExprKind::MethodCall(method, conditional_receiver, _, _) = unary_inner.kind
+            && method.ident.as_str() == "is_empty"
+            && let expr_ty = cx.typeck_results().expr_ty(conditional_receiver).peel_refs()
+            && (expr_ty.is_array_slice() || expr_ty.is_array() || is_type_diagnostic_item(cx, expr_ty, sym::Vec))
+            && let ExprKind::Block(block, _) = if_expr.then.kind
+        {
+            // the receiver for the index operation
+            let mut index_receiver: Option<&Expr<'_>> = None;
+            // first local in the block - used as pattern for `Some(pat)`
+            let mut first_local: Option<&Local<'_>> = None;
+            // any other locals to be aware of, these are set to the value of `pat`
+            let mut extra_locals: Vec<&Local<'_>> = vec![];
+            // any other index expressions to replace with `pat` (or "element" if no local exists)
+            let mut extra_exprs: Vec<&Expr<'_>> = vec![];
+
+            for_each_expr(block.stmts, |x| {
+                if let ExprKind::Index(receiver, index, _) = x.kind
+                    && let ExprKind::Lit(lit) = index.kind
+                    && let LitKind::Int(val, _) = lit.node
+                    && eq_expr_value(cx, receiver, conditional_receiver)
+                    && val.0 == 0
+                {
+                    index_receiver = Some(receiver);
+                    if let Node::Local(local) = cx.tcx.parent_hir_node(x.hir_id) {
+                        if first_local.is_none() {
+                            first_local = Some(local);
+                        } else {
+                            extra_locals.push(local);
+                        };
+                    } else {
+                        extra_exprs.push(x);
+                    };
+                } else if let ExprKind::AddrOf(BorrowKind::Ref, Mutability::Mut, val) = x.kind
```

Comment:
There's also auto borrows to consider e.g.

```rust
let mut v = vec![1, 2, 3];
if !v.is_empty() {
    v.push(1);
    let b = v[0];
    b;
}
```

You can look through the [`expr_adjuments`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.expr_adjustments) of the slice/vec to see if any are `Borrow` with `AutoBorrowMutability::Mut`

Reply:
Trickier is when the thing we're linting is a mutable reference to begin with as there's no adjustment needed, it might be worth ignoring these for now unless you can come up with a nice way to handle them

```rust
let v = &mut vec![1, 2, 3];
if !v.is_empty() {
    let b = v[0];
    v; // or v.push(1), etc.
    b;
}
```

Reply:
Also not hugely likely to occur but this will also catch [`addr_of_mut`](https://doc.rust-lang.org/std/ptr/macro.addr_of_mut.html)

```suggestion
                } else if let ExprKind::AddrOf(_, Mutability::Mut, val) = x.kind
```